### DISCUSSION TITLE
Direct Control Loop Rate

### DISCUSSION THREAD

**Discussion by nicholasforslund** [battery-data-alliance/pymacnet#14] [Q&A]
I am interested in using the direct control mode instead of defining test procedures so I can use a different test sequence environment. Has anyone benchmarked the loop rate in the direct control mode? I would like to understand at least an order of magnitude for how fast the direct control mode can run. Any tips or experience with performance in direct control mode is much appreciated!

**Comment by chintanp** [comment]
@nicholasforslund while we havent published any benchmarks from our testing, last we heard from Maccor, this interval was 100 ms, i.e. this is how fast the cycler software would react to external commands. What refresh rate are you expecting for your usecase?

**Reply by nicholasforslund** [reply]
@chintanp Ideally, control would be possible at 100 Hz (or even up to 1000 Hz), but I understand that this is limited by the response from MacNet. Is this 100 ms response rate for all commands? I am also exploring running a test procedure and updating those variables dynamically if that would increase performance.

**Reply by chintanp** [reply]
I am trying to understand which tests would need such a high control rate. Regardless, 100 ms seems to be the best for all commands. Having a test procedure and updating vars is called "Variable Control Mode" instead of "Direct Control Mode" where no test procedure is used. Either case has a max write rate of 100 ms.
